### DISCUSSION TITLE
unwrap_intf.csh

### DISCUSSION THREAD

**Discussion by galuhqfs** [gmtsar/gmtsar#404] [Q&A]
Hi,

Now, I running unwrap_intf.csh for my timeseries data. However, it takes so long to end the unwrap_intf.csh process. It toke a week and it doesn't seems to come to an end process. 

galuhqoriah@DESKTOP-U2QNNPU:/mnt/f/Skripsi/Data/Cimandiri2/asc/merge2$ unwrap_intf.csh 2020061_2020097
corr_patch.grd mask_def_patch.grd MUL = corr_patch.grdSetting search radius to be 300 ...
Reading in original grid...
Interpolating to nearest neighbour...
Working on line 33 ...

The working on line is still progressing. Before this, I tried to cancel the process when it was already on line 10000, and it still doesn't come to an end. Is it normal for unwrapping takes longer than any process before?
Thank you

**Comment by Xiaohua-Eric-Xu** [comment]
The nearneighbor interpolation in the package is very slow when the image has a large blank area such as ocean, and is very fast for interpolating through decorrelation like from snow. If interpolation is not necessary for your area, you should consider using snaphu.csh instead.

**Reply by galuhqfs** [reply]
I see.

Anyway, I have another question regarding to unwrapping process. According to your sentinel_time_series_3.pdf tutorial, the command that going to use is `snaphu_interp.csh 0.10 40 1000/3000/24000/27000`. What's the meaning of number 0.10 and 40,  by any chance?
Thank you.

**Reply by Xiaohua-Eric-Xu** [reply]
Please refer to the newest version of the document. http://topex.ucsd.edu/gmtsar/tar/sentinel_time_series_5.pdf

As for this specific command, 0.10 is the coherence threshold, i.e. pixels with coherence smaller than that will be masked out.
And 40 is the maximum discontinuities you allow to exist after unwrapping (in terms of number of phase cycles).

**Reply by galuhqfs** [reply]
I have another question for you. Where do I have to find out the number of  minRange/maxRange/minAzi/maxAzi?

When I runned intf process, the result showed a .pdf file that given an information about phase, range, and azimuth. Nevertheless, on merge process there aren't any file with .pdf format.

**Reply by Xiaohua-Eric-Xu** [reply]
I think I didn't write the plot part in the merging script. It's basically a simple plotting with `gmt`. There is a handy program called `ncview` that can get you a quick look on the netcdf files. I usually use that instead of really making a plot.

**Reply by galuhqfs** [reply]
I got an error while running the `snaphu_interp.csh 0.10 40 12330/42750/8026/12706`. Here is the result:

grdcut [ERROR]: Error for input file: No such file (mask.grd)
grdcut [ERROR]: Error for input file: No such file (corr.grd)
grdcut [ERROR]: Error for input file: No such file (phasefilt.grd)
grdmath [ERROR]: Syntax error: corr_patch.grd is not a number, operator or file name
grdmath [ERROR]: Syntax error: corr_patch.grd is not a number, operator or file name
grdmath [ERROR]: Syntax error: corr_patch.grd is not a number, operator or file name
grdmath [ERROR]: Syntax error: mask2_patch.grd is not a number, operator or file name
grdmath [ERROR]: Syntax error: mask2_patch.grd is not a number, operator or file name
Setting search radius to be 300 ...
Reading in original grid...
nearest_grid [ERROR]: File phase_tmp.grd not found
[Session nearest_grid (0)]: Error returned from GMT API: GMT_FILE_NOT_FOUND (16)
 cannot open grdfile phase_tmp.grd
mv: cannot stat 'tmp.grd': No such file or directory
grd2xyz [ERROR]: File phase_tmp.grd not found
[Session gmt (0)]: Error returned from GMT API: GMT_FILE_NOT_FOUND (16)
grd2xyz [ERROR]: File corr_tmp.grd not found
[Session gmt (0)]: Error returned from GMT API: GMT_FILE_NOT_FOUND (16)
unwrapping phase with snaphu - higher threshold for faster unwrapping
grdinfo [ERROR]: Error for input file: No such file (phase_patch.grd)
grdinfo [ERROR]: Syntax error: Must specify one or more input files

snaphu v2.0.1
32 parameters input from file snaphu.conf.brief (297 lines total)
not enough input arguments.  type snaphu -h for help
grdinfo [ERROR]: Error for input file: No such file (phase_patch.grd)
grdinfo [ERROR]: Syntax error: Must specify one or more input files
grdinfo [ERROR]: Error for input file: No such file (phase_patch.grd)
grdinfo [ERROR]: Syntax error: Must specify one or more input files
xyz2grd [ERROR]: Error for input file: No such file (unwrap.out)
xyz2grd [ERROR]: Syntax error: Must specify -R option
xyz2grd [ERROR]: Syntax error -I option: Must specify positive increment(s)
grdinfo [ERROR]: Error for input file: No such file (phase_patch.grd)
grdinfo [ERROR]: Syntax error: Must specify one or more input files
grdinfo [ERROR]: Error for input file: No such file (phase_patch.grd)
grdinfo [ERROR]: Syntax error: Must specify one or more input files
xyz2grd [ERROR]: Error for input file: No such file (conncomp.out)
xyz2grd [ERROR]: Syntax error: Must specify -R option
xyz2grd [ERROR]: Syntax error -I option: Must specify positive increment(s)
grdmath [ERROR]: Syntax error: tmp.grd is not a number, operator or file name
mv: cannot stat 'tmp.grd': No such file or directory
grdmath [ERROR]: Syntax error: unwrap.grd is not a number, operator or file name
mv: cannot stat 'tmp.grd': No such file or directory
grdgradient [ERROR]: Error for input file: No such file (unwrap.grd)
grdgradient [ERROR]: Syntax error: Must specify input file
grdinfo [ERROR]: Error for input file: No such file (unwrap.grd)
grdinfo [ERROR]: Syntax error: Must specify one or more input files
makecpt [ERROR]: Syntax error -T: Must specify valid min[/max[/inc[<unit>|+n]]] option
grdinfo [ERROR]: Error for input file: No such file (unwrap.grd)
grdinfo [ERROR]: Syntax error: Must specify one or more input files
grdinfo [ERROR]: Error for input file: No such file (unwrap.grd)
grdinfo [ERROR]: Syntax error: Must specify one or more input files
grdimage [ERROR]: Error for input file: No such file (unwrap.grd)
grdimage [ERROR]: Syntax error -I: Requires a valid grid file or a constant
grdimage [ERROR]: Syntax error: Must specify one (or three) input file(s)
grdimage [ERROR]: Syntax error -I option: Must specify intensity file, value, or modifiers
psscale [ERROR]: Could not parse nwrap.grd into Geographical, Cartesian, or Temporal coordinates!psscale [ERROR]: Syntax error -R option. Correct syntax:
        -R<xmin>/<xmax>/<ymin>/<ymax>[/<zmin>/<zmax>]
          Append r if giving lower left and upper right coordinates
        -Rg or -Rd for global domain
        -R<grdfile> to take the domain from a grid file
psscale [ERROR]: Offending option -Runwrap.grd
psconvert [ERROR]: The file unwrap.ps has no BoundingBox in the first 20 lines or last 256 bytes. Use -A option.
Unwrapped phase map: unwrap.pdf
mv: cannot stat 'phase_patch.grd': No such file or directory

From that error, what's wrong with my process?

**Reply by Xiaohua-Eric-Xu** [reply]
> grdcut [ERROR]: Error for input file: No such file (mask.grd)

This indicate you did not run the script at the right directory. Three files are needed, mask.grd, corr.grd and phasefilt.grd

**Reply by galuhqfs** [reply]
Hi, I have a question related to unwrapping process using `snaphu_interp.csh`. I have 33 pairs of interferogram that need to be unwrapped. All process went well, until I realized that one of the interferogram doesn't have the same dimension. The corr.grd file of 2020229_2020253 folder didn't cut the region area. I used the same script to run the snaphu, the script : `snaphu_interp.csh 0.10 40 12330/42750/8026/12706`.

Here is the information of my corr.grd file before I run the snaphu:
galuhqoriah@DESKTOP-U2QNNPU:/mnt/f/Skripsi/Data/Cimandiri2/asc/merge3/2020229_2020253$ gmt grdinfo corr.grd
corr.grd: Title: merged grid
corr.grd: Command:
corr.grd: Remark:
corr.grd: Pixel node registration used [Cartesian grid]
corr.grd: Grid file format: nf = GMT netCDF format (32-bit float), CF-1.7
corr.grd: x_min: 0 x_max: 45440 x_inc: 4 name:  n_columns: 11360
corr.grd: y_min: 0 y_max: 13989 y_inc: 1 name:  n_rows: 13989
corr.grd: z_min: 0.00298334495164 z_max: 0.993842244148 name: z
corr.grd: scale_factor: 1 add_offset: 0
corr.grd: format: netCDF-4 chunk_size: 130,129 shuffle: on deflation_level: 3

And this is the result of snaphu process:
 galuhqoriah@DESKTOP-U2QNNPU:/mnt/f/Skripsi/Data/Cimandiri2/asc/merge3/2020229_2020253$ snaphu_interp.csh 0.10 40 12330/42750/8026/12706
Setting search radius to be 300 ...
Reading in original grid...
Interpolating to nearest neighbour...
Working on line 4680 ...
  -68662237 number of searches used ...
WRITING GRID IMAGE: Width x Heihgt = 7606 x 4680...
unwrapping phase with snaphu - higher threshold for faster unwrapping

snaphu v2.0.1
32 parameters input from file snaphu.conf.brief (297 lines total)
Reading wrapped phase from file phase.in
No weight file specified.  Assuming uniform weights
Reading correlation data from file corr.in
Calculating deformation-mode cost parameters
Building range cost arrays
Building azimuth cost arrays
Initializing flows with MST algorithm
Running approximate minimum spanning tree solver
Running nonlinear network flow optimizer
Maximum flow on network: 11
Number of nodes in network: 35583796
Flow increment: 1  (Total improvements: 0)
Found 1 valid set(s) of connected pixels
Source 0 row, col = -2, -2
Treesize: 35583796   Pivots: 479983      Improvements: 4842
Maximum valid flow on network: 8
Flow increment: 2  (Total improvements: 4842)
Found 1 valid set(s) of connected pixels
Source 0 row, col = -2, -2
Treesize: 35583796   Pivots: 5918        Improvements: 261
Maximum valid flow on network: 8
Flow increment: 3  (Total improvements: 5103)
Found 1 valid set(s) of connected pixels
Source 0 row, col = -2, -2
Treesize: 35583796   Pivots: 1169        Improvements: 58
Maximum valid flow on network: 9
Flow increment: 4  (Total improvements: 5161)
Found 1 valid set(s) of connected pixels
Source 0 row, col = -2, -2
Treesize: 35583796   Pivots: 430         Improvements: 16
Growing connected component mask
1 connected components formed
Writing connected components to file conncomp.out as 1-byte unsigned ints
Maximum flow on network: 8
Total solution cost: 32750865
Integrating phase
Writing output to file unwrap.out
Program snaphu done
Elapsed processor time:   0:08:43.80
Elapsed wall clock time:  0:12:29
Unwrapped phase map: unwrap.pdf

And the dimension of corr.grd doesn't change
galuhqoriah@DESKTOP-U2QNNPU:/mnt/f/Skripsi/Data/Cimandiri2/asc/merge3/2020229_2020253$ gmt grdinfo corr.grd
corr.grd: Title: merged grid
corr.grd: Command:
corr.grd: Remark:
corr.grd: Pixel node registration used [Cartesian grid]
corr.grd: Grid file format: nf = GMT netCDF format (32-bit float), CF-1.7
corr.grd: x_min: 0 x_max: 45440 x_inc: 4 name:  n_columns: 11360
corr.grd: y_min: 0 y_max: 13989 y_inc: 1 name:  n_rows: 13989
corr.grd: z_min: 0.00298334495164 z_max: 0.993842244148 name: z
corr.grd: scale_factor: 1 add_offset: 0
corr.grd: format: netCDF-4 chunk_size: 130,129 shuffle: on deflation_level: 3

**Reply by Xiaohua-Eric-Xu** [reply]
The script will not touch corr.grd. It'll only produce a unwrap.grd with the cut parameter. If you want to cut your corr.grd, use
```
gmt grdcut corr.grd -R12330/42750/8026/12706 -Gcorr_cut.grd
```

**Reply by galuhqfs** [reply]
Thank you so much for your help. All of my problem on unwrapping process has been solved.

Moving on to the next process is SBAS. I have some issues regarding to rms misfit and dem error. I did some research and found that issue cause by memory allocated. And it can be solved by using `-mmap`. But I have no idea where to put the `-mmap` and I know nothing about `-mmap` itself. Could you give me a clue, by any chance?

Thank you in advance.

**Reply by Xiaohua-Eric-Xu** [reply]
`sbas balabla_with_all_your_commands -mmap`
mmap swaps disk storage to be used as memory. It may slow the process down by a bit but grants you much more fake memory.

**Reply by galuhqfs** [reply]
Thank you so much for your help, Eric.

**Reply by galuhqfs** [reply]
Hi, Eric. I have a question regarding to the unwrapping process using or not using mask in the process itself. 

If we use `unwrap_intf.csh` for unwrap our intf, the csh file has mention mask_def.grd file. So we include the mask for unwrap the intf. Furthermore, the `snaphu_interp.csh` hasnt mention the mask_def.grd at all.

And my understanding is that the unwrap_intf.csh will have a better result than the snaphu_interp.csh. But, in my case, I tried to run all the command and found that my unwrap data has similar yet different. For the water area, the snaphu_interp.csh has better result.

My question is, is it really that necesarry for me to mask the phase? And what's the correlation for using and not using mask for unwrapping process?

Thank you

**Reply by Xiaohua-Eric-Xu** [reply]
Yeah, that's a hidden part in the program. The script will search for the file given that specific name. It is not an ideal way of doing stuff. Similar case is landmask_ra.grd. 

I am not sure why there's a difference, if your unwrap_intf.csh is calling snaphu_interp.csh. Or maybe yours calls snaphu.csh instead?

The masking is really an arbitrary thing to do. Like picking a 0.1 threshold is not that much different from picking 0.11 (I usually use 0.1 unless you have other mask files activated e.g. mask_def.grd). It really depends on what you wanna get from the data.

**Reply by galuhqfs** [reply]
Thank you so much for your help
